content/articles/ioc_principle/index.md

```diff
@@ -0,0 +1,250 @@
+---
+layout: engineering-education
+status: publish
+published: true 
+url: /engineering-education/inversion_of_control_principle/
+title: Inversion of control principle
+description: Inversion of Control(IoC) is a software design principle used in object-oriented programming to reverse control and achieve loose coupling of classes. Coupling measures how closely connected software modules are to each other, while control refers to any other additional responsibility a class has apart from its primary responsibility. Remember the S in SOLID principles of object-oriented design? Let me help you: Singularity of Purpose—that is the objective of IoC. Loose coupling in software engineering is often a sign of testable, well-structured, maintainable, and readable software.
+author: love-otudor
+date: 2021-07-16T00:00:00-11:00
+topics: [Languages]
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/inversion_of_control_principle/hero.jpeg
+    alt: Inversion of control principle cover image 
+---
+
+Inversion of Control(IoC) is a software design principle. It helps to reverse control in object-oriented programming. Control here is any extra responsibility a class has besides its main responsibility. IoC passes the control of an object’s creation, the flow of an application, etc. to a container or framework.
+
+IoC inverts the control of creating and managing objects from the programmer to a container. This container manages the creation and lifecycle of objects in the application. IoC helps us create large systems by eradicating the responsibility of creating objects.
+
+Imagine this: suppose you have a bakery that specializes in making egg rolls. What is an egg roll without an egg? Right. So eggs are essential in the making of egg rolls. Why egg rolls you may ask? Well, you could call it plain nepotism. So how do we get a bakery that obeys the IoC principle? Well, rather than having a poultry farm and involving yourself with all the poultry-related tasks, you order the eggs and get them delivered to you. This time, you don't have to worry about the chicks and the eggs being healthy, or if you are using the right feed, and all. With this new design, you can focus on making egg rolls, which results in a more effective baking process. This is the big idea behind IoC: separating concerns.
+
+### Here are some of the benefits of using IoC:
+* Easy to switch between different implementations of a particular class at runtime.
+* It increases the modularity of the program.
+* It manages an object’s lifecycle and configuration. Some objects can be singletons, while we can create others per request.
+
+Before we look at some code, here are some things you should have in mind:
+
+### 1. Coupling
```

Comment:
I think we can also have the advantages of this just like you have done in step three.